lib/auth/auth_with_roles.go

```diff
@@ -2005,6 +2029,84 @@ func (a *ServerWithRoles) ProcessKubeCSR(req KubeCSR) (*KubeCSRResponse, error)
 	return a.authServer.ProcessKubeCSR(req)
 }
 
+// GetDatabaseServers returns all registered database servers.
+func (a *ServerWithRoles) GetDatabaseServers(ctx context.Context, namespace string, opts ...services.MarshalOption) ([]services.DatabaseServer, error) {
+	if err := a.action(namespace, services.KindDatabaseServer, services.VerbList); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := a.action(namespace, services.KindDatabaseServer, services.VerbRead); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	servers, err := a.authServer.GetDatabaseServers(ctx, namespace, opts...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// Filter out databases the caller doesn't have access to from each server.
+	var filtered []services.DatabaseServer
+	for _, server := range servers {
+		err := a.context.Checker.CheckAccessToDatabaseServer(server)
+		if err != nil && !trace.IsAccessDenied(err) {
+			return nil, trace.Wrap(err)
+		} else if err == nil {
+			filtered = append(filtered, server)
+		}
+	}
+	return filtered, nil
+}
+
+// UpsertDatabaseServer creates or updates a new database proxy server.
+func (a *ServerWithRoles) UpsertDatabaseServer(ctx context.Context, server services.DatabaseServer) (*services.KeepAlive, error) {
+	if err := a.action(server.GetNamespace(), services.KindDatabaseServer, services.VerbCreate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := a.action(server.GetNamespace(), services.KindDatabaseServer, services.VerbUpdate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return a.authServer.UpsertDatabaseServer(ctx, server)
+}
+
+// DeleteDatabaseServer removes the specified database proxy server.
+func (a *ServerWithRoles) DeleteDatabaseServer(ctx context.Context, namespace, hostID, name string) error {
+	if err := a.action(namespace, services.KindDatabaseServer, services.VerbDelete); err != nil {
+		return trace.Wrap(err)
+	}
+	return a.authServer.DeleteDatabaseServer(ctx, namespace, hostID, name)
+}
+
+// DeleteAllDatabaseServers removes all registered database proxy servers.
+func (a *ServerWithRoles) DeleteAllDatabaseServers(ctx context.Context, namespace string) error {
+	if err := a.action(namespace, services.KindDatabaseServer, services.VerbList); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := a.action(namespace, services.KindDatabaseServer, services.VerbDelete); err != nil {
+		return trace.Wrap(err)
+	}
+	return a.authServer.DeleteAllDatabaseServers(ctx, namespace)
+}
```

Comment:
Hm, interesting question, I'm not sure why it would be necessary here and resource/action checks should be enough? It also seems to be consistent with how similar methods for Kubernetes/Apps are implemented here as well.

Reply:
my thinking was: if you can't access/list database "foo", you shouldn't be able to update or delete it.
I think I had the same discussion in the AAP PR and was overruled.

let's stick with consistency for now.
in the future, when our API gets used more externally, we may need to patch this part of RBAC behavior.